### DISCUSSION TITLE
confusing by rules %value%

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22503] [Q&A]
if I do check rules docu
`rule1 ON LD2410#Distance DO %value% ENDON`
 should shown anything with the value of trigger in the output.
But it is empty each time. is that caused by var being a list ?

```
14:25:50.848 CMD: rule1 ON LD2410 DO %value% ENDON
14:25:50.854 RUL: Stored uncompressed, would compress from 26 to 24 (-8%)
14:25:50.857 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":26,"Free":485,"Rules":"ON LD2410 DO %value% ENDON"}}
14:25:51.071 RUL: LD2410 performs ''
14:25:52.073 RUL: LD2410 performs ''
14:25:53.171 RUL: LD2410 performs ''
14:25:54.073 RUL: LD2410 performs ''
14:25:54.170 CMD: rule1 ON LD2410#Distance DO %value% ENDON
14:25:54.175 RUL: Stored uncompressed, would compress from 35 to 29 (-17%)
14:25:54.179 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"ON LD2410#Distance DO %value% ENDON"}}
14:25:55.068 RUL: LD2410#DISTANCE performs ''
14:25:56.169 RUL: LD2410#DISTANCE performs ''
14:25:57.074 RUL: LD2410#DISTANCE performs ''
14:25:58.160 RUL: LD2410#DISTANCE performs ''
14:25:59.163 RUL: LD2410#DISTANCE performs ''
14:26:00.074 RUL: LD2410#DISTANCE performs ''
14:26:00.265 RSL: SENSOR = {"Time":"2024-11-17T14:26:00","Switch8":"OFF","ANALOG":{"CTEnergy":{"Energy":0.000,"Power":0,"Voltage":230,"Current":0.000}},"LD2410":{"Distance":[88.0,88.0,78.0],"Energy":[0,93]}}
14:26:01.180 RUL: LD2410#DISTANCE performs ''
14:26:04.412 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"ON LD2410#Distance DO %value% ENDON"}}
```

**Comment by sfromis** [comment]
No, a syntax error has no "should" for what happens. To see what values you get, it works better with something like: 
`ON LD2410#Distance DO var1 %value% ENDON`
(Some types of syntax errors will not even cause the data to be seen as a command)

Also, using rules is not at all great for arrays, but you can still pick individual fields from arrays like this:
`on LD2410#Distance[1] do var1 %value% endon`

**Reply by ChrisVonFrosta** [reply]
since i read values 0-500 and the dimmer needs 0-100. can I norm it through calculating inside the rule?
somthing like that?: 
`rule1 ON LD2410#Distance[1] DO 500-((%value%/500)*100) ENDON`
or
`rule1 ON LD2410#Distance[1] DO var1 %value%/500*100 ENDON ON LD2410#Distance[1] DO 500 - %var1% ENDON`

**Reply by sfromis** [reply]
No to both.

You cannot just have an expression and call it a rule, you need to assign to a variable.

And if you want to have an assignment expression, you need something like `var1=%value%/500*100` - without that **=** it is no expression to be evaluated.

Apart from this, I see neither of your expressions making sense if what you want is value from 0 to 100. What could work for this would be a simple divide by 5 to get from 0..500 to 0..100, or with more flexibility for tuning endpoints, you can set `var2` with a command like 
`Scale2 %value%,0,500,0,100`

In all cases, you may want to config `CalcRes 0` first, to have zero decimals on the result to be used with the `Dimmer` command.

**Reply by ChrisVonFrosta** [reply]
wow. so nice from you! thank for teaching

**Reply by ChrisVonFrosta** [reply]
>I see neither of your expressions making sense

it is just for experimental with LD2410. i want to get brighter light when coming closer

**Reply by ChrisVonFrosta** [reply]
i end without success. i tried all variants.

`Scale2 %value%,0,500,0,100` works but i can't store the result to var1.

`Rule2 ON STATUS DO var1 Scale2 77,0,500,0,100 ENDON DO Dimmer %var1% ENDON` does `STATUS performs 'var1 Scale2 77,0,500,0,100'` `RESULT = {"Var1":"Scale2 77,0,500,0,100"}`

`Rule2 ON STATUS DO var1=Scale2 77,0,500,0,100 ENDON DO Dimmer %var1% ENDON` does `STATUS performs 'var1=Scale2 77,0,500,0,100'`  `RESULT = {"Var1":"0"}`

**Reply by sfromis** [reply]
If you want it in var1, use `Scale1` instead of `Scale2`, but do not prefix the `Scale` command by a `Var` command. As indicated in what I posted, `Scale` is a **command**, not some function/operator for the `Var` command.
